### PR TITLE
refactored the auth and routes

### PR DESCRIPTION
**Pull request by nitzantomer**

### CODE REVIEW

File: scripts/src/index.ts [hunk 1]
```diff
@@ -55,7 +54,7 @@ app.use((req, res, next) => {
 // catch errors
 app.use((err, req, res, next) => {
 	const status = err.status || 500;
-	// log.error(`Error ${status} (${err.message}) on ${req.method} ${req.url} with payload ${req.body}.`);
+	logger.error(`Error ${status} (reason: ${err.message}) on ${req.method} ${req.url} with payload ${req.body}.`);
```
Comment (doodyparizada) [inline scripts/src/index.ts]: add stack trace

File: scripts/src/middleware.ts [hunk 2]
```diff
@@ -40,4 +50,4 @@ export async function userContext(
 	}
 	req.context = { user, authToken };
 	next();
-}
+}*/
```
Comment (doodyparizada) [inline scripts/src/middleware.ts]: why is all of this commented out? delete if not needed

File: scripts/src/routes/index.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+function router(): ExtendedRouter {
+	const router = express.Router() as ExtendedRouter;
+
+	router.authenticated = function() {
+		return new Proxy(this, {
+			get(target, name) {
+				if (typeof name === "string" && AUTHENTICATED_METHODS.includes(name)) {
+					return (path: string, handler: express.RequestHandler) => {
+						return target[name](path, async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+							const token = await checkAuthentication(req);
```
Comment (doodyparizada) [inline scripts/src/routes/index.ts]: maybe call it `authenticate()` check sounds like a void function

File: scripts/src/routes/index.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		router()
+			.get("/", getUser)
+			.authenticated()
+				.post("/", signinUser)
+				.post("/me/activate", activateUser));
+}
+
+function createPath(path: string, prefix?: string): string {
+	if (!prefix) {
+		return prefix;
```
Comment (doodyparizada) [inline scripts/src/routes/index.ts]: you mean `return path`

File: scripts/src/routes/offers.ts [hunk 5]
```diff
@@ -1,53 +1,48 @@
...
-	// user_offers = []
-	// for offer in offers:
-	//   if len(OpenOrder.find({offerId} && not userId)) + offer.used >= offer.cap
-	//   or len(Transaction.find({offerId, userId})) >= offer.user_cap:
-	//     user_offers += offer
-	// return user_offers
+export async function getOffers(req: Request, res, next) {
 	try {
-		const data: OfferList = await getOffers(req.context.user.id, req.context.user.appId);
+		const data = await getOffersService(req.context.user.id, req.context.user.appId);
```
Comment (doodyparizada) [inline scripts/src/routes/offers.ts]: why renamed to Service?

File: scripts/src/routes/router.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+		token: string;
+		context: Context;
+	}
+}
+
+export type ExtendedRouter = Router & {
+	authenticated(): Router;
+};
+
+const AUTHENTICATED_METHODS = ["get", "delete", "post", "put", "patch"];
```
Comment (doodyparizada) [inline scripts/src/routes/router.ts]: this code is in both `index.ts` and `router.ts`

File: scripts/src/routes/offers.ts (hunk 5 above)
Comment (nitzantomer) [reply scripts/src/routes/offers.ts]: cuz i like it this way!

Comment (doodyparizada) [reply scripts/src/routes/offers.ts]: well you don't get a pokemon go

File: scripts/src/auth.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import * as express from "express";
+
+import * as db from "./models/users";
+
+export async function checkAuthentication(req: express.Request): Promise<db.AuthToken> {
```
Comment (doodyparizada) [inline scripts/src/auth.ts]: it's still called checkauthentiocfaaijtiuihn

**Pull request merged**